plugin/src/main/java/org/opensearch/ml/action/models/UpdateModelTransportAction.java

```diff
@@ -102,66 +119,79 @@ protected void doExecute(Task task, ActionRequest request, ActionListener<Update
         String[] excludes = new String[] { MLModel.MODEL_CONTENT_FIELD, MLModel.OLD_MODEL_CONTENT_FIELD };
 
         try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+            ActionListener<UpdateResponse> wrappedListener = ActionListener.runBefore(actionListener, context::restore);
             mlModelManager.getModel(modelId, null, excludes, ActionListener.wrap(mlModel -> {
-                FunctionName functionName = mlModel.getAlgorithm();
-                MLModelState mlModelState = mlModel.getModelState();
-
-                if (functionName == TEXT_EMBEDDING || functionName == REMOTE) {
-                    if (mlModel.getIsHidden() != null && mlModel.getIsHidden()) {
-                        if (isSuperAdmin) {
-                            updateRemoteOrTextEmbeddingModel(modelId, updateModelInput, mlModel, user, actionListener, context);
-                        } else {
-                            actionListener
-                                .onFailure(
-                                    new OpenSearchStatusException(
-                                        "User doesn't have privilege to perform this operation on this model, model ID " + modelId,
-                                        RestStatus.FORBIDDEN
-                                    )
+                if (!isModelDeploying(mlModel.getModelState())) {
+                    boolean isModelDeployed = isModelDeployed(mlModel.getModelState());
+                    FunctionName functionName = mlModel.getAlgorithm();
+                    if (functionName == TEXT_EMBEDDING || functionName == REMOTE) {
+                        if (mlModel.getIsHidden() != null && mlModel.getIsHidden()) {
+                            if (isSuperAdmin) {
+                                updateRemoteOrTextEmbeddingModel(
```

Comment:
Why only support updating remote or text embedding model?

Reply:
This is by design. Built-in functions and other new functions(metrics correlation, sparse matrix, dense matrix) need too much additional efforts at this time.

Reply:
I think we've discussed this in the design review meeting.

Reply:
Suggest add some comments to code , so others can easily understand the reason and limitation.

Reply:
We recently added another [algorithm](https://github.com/opensearch-project/ml-commons/blob/01cf00c2e86fb09e41ff95cae16e4e1a0d29ab0a/common/src/main/java/org/opensearch/ml/common/FunctionName.java#L23) TEXT SIMILARITY, which should be similar to TEXT EMBEDDING.

We don't need to support update TEXT SIMILARITY in this PR, but let's add a `todo` to come back later in these use cases.

Reply:
Added